### PR TITLE
feat: add global labels to service metrics

### PR DESCRIPTION
**Pull request by kruskall**
## Motivation/summary

Add labels to service metrics struct and move comparable fields to a separate struct.
The new struct is not comparable and cannot be used as a map key so replace the map with a custom hash map using separate chaining for collision resolution and implement a custom hash function.

Add systemtest to verify global labels are working as intended.

## Checklist

- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)
- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)
- [ ] Documentation has been updated

For functional changes, consider:
- Is it observable through the addition of either **logging** or **metrics**?
- Is its use being published in **telemetry** to enable product improvement?
- Have system tests been added to avoid regression?

## Related issues

Related to https://github.com/elastic/apm-server/issues/9060

### CODE REVIEW

**Review by marclop — COMMENTED** [review]
Implementation looks good. I think we can refactor some of the code to avoid duplication and better reuse what already exists.

File: x-pack/apm-server/aggregation/servicemetrics/aggregator.go [hunk 1]
```diff
@@ -252,15 +320,49 @@ type aggregationKey struct {
...
+		}
+		if len(v.Values) > 0 {
+			key.numericLabels.SetSlice(k, v.Values)
+		} else {
+			key.numericLabels.Set(k, v.Value)
+		}
+		key.numericLabelKeys = append(key.numericLabelKeys, k)
 	}
+	sort.Strings(key.labelKeys)
+	sort.Strings(key.numericLabelKeys)
```
[the remark this answers is not in the document]
Comment (kruskall) [reply x-pack/apm-server/aggregation/servicemetrics/aggregator.go]: Great point! Moved to another package!

File: x-pack/apm-server/aggregation/servicemetrics/aggregator.go [hunk 2]
```diff
@@ -312,3 +416,86 @@ func makeMetricset(key aggregationKey, metrics serviceMetrics) model.APMEvent {
...
+			return false
+		}
+		for i, value := range v.Values {
+			if localV.Values[i] != value {
+				return false
+			}
+		}
+	}
+	return true
+}
```
[the remark this answers is not in the document]
Comment (kruskall) [reply x-pack/apm-server/aggregation/servicemetrics/aggregator.go]: Agree! Fixed :+1:

File: x-pack/apm-server/aggregation/servicemetrics/aggregator.go [hunk 3]
```diff
@@ -204,45 +210,107 @@ func (a *Aggregator) processTransaction(event *model.APMEvent) model.APMEvent {
...
+	mb.m[hash] = append(entries, entry)
+	mb.entries++
 	return true
 }
 
 type aggregationKey struct {
+	labelKeys        []string
+	labels           model.Labels
+	numericLabelKeys []string
+	numericLabels    model.NumericLabels
```
[the remark this answers is not in the document]
Comment (kruskall) [reply x-pack/apm-server/aggregation/servicemetrics/aggregator.go]: I moved the global labels to a separate package, let me know what you think!

File: x-pack/apm-server/aggregation/servicemetrics/aggregator.go [hunk 4]
```diff
@@ -204,45 +209,101 @@ func (a *Aggregator) processTransaction(event *model.APMEvent) model.APMEvent {
...
+	k.AggregatedGlobalLabels.Write(&h)
+	h.WriteString(k.agentName)
+	h.WriteString(k.serviceEnvironment)
+	h.WriteString(k.serviceName)
+	h.WriteString(k.transactionType)
+	return h.Sum64()
+}
+
+func (k *aggregationKey) equal(key aggregationKey) bool {
+	return k.comparable == key.comparable && k.AggregatedGlobalLabels.Equals(&key.AggregatedGlobalLabels)
```
Comment (marclop) [inline x-pack/apm-server/aggregation/servicemetrics/aggregator.go:303]: nit; I think this reads slightly better on smaller editors.
```suggestion
	return k.comparable == key.comparable && 
		k.AggregatedGlobalLabels.Equals(&key.AggregatedGlobalLabels)
```

File: x-pack/apm-server/aggregation/txmetrics/aggregator.go [hunk 5]
```diff
@@ -612,9 +579,7 @@ func (k *transactionAggregationKey) hash() uint64 {
 }
 
 func (k *transactionAggregationKey) equal(key transactionAggregationKey) bool {
-	return k.comparable == key.comparable &&
-		equalLabels(k.labels, key.labels) &&
-		equalNumericLabels(k.numericLabels, key.numericLabels)
+	return k.comparable == key.comparable && k.AggregatedGlobalLabels.Equals(&key.AggregatedGlobalLabels)
```
Comment (marclop) [inline x-pack/apm-server/aggregation/txmetrics/aggregator.go:582]:
```suggestion
	return k.comparable == key.comparable &&
		k.AggregatedGlobalLabels.Equals(&key.AggregatedGlobalLabels)
```

**Review by marclop — APPROVED** [review]
Looks good!

**Comment by marclop** [comment]
I forgot during the review, but we should add a changelog entry for this change as well

**Comment by axw** [comment]
Tracking test-plan status in https://github.com/elastic/apm-server/issues/9060

**Pull request merged**